File: src2/notesExplorer.js
import {
  getClosestRatioFromScale,
  computeRotations,
  unique,
  ratioToCents,
  getMinInterval
} from "./MicrotonalHelper/operations.js";
import * as mos from "./MicrotonalHelper/mos.js";
import * as fs from "fs";
import {
  getEqualTemperamentNote,
  getEqualTemperamentSubsets
} from "./MicrotonalHelper/equalTemperament.js";
import {
  diamond5Limit,
  diamond7Limit,
  diamond9Limit,
  diamond9LimitNo7
} from "./MicrotonalHelper/diamond.js";

//Notes Explorer
//===========================
//This searches for scales of N notes that satisfy the target ratios considering all related modes (rotations)
//---------------------------
//Parameters:
const numberOfNotes = 6;
const numberOfDivisions = 22;
const period = 2;
const maxToleranceInCents = 16;
const noteDistance = 1; //numberOfDivisions > 43 ? 2 : 1;
const targetRatios = [
  //, 10 / 5, 10 / 4, 8 / 3 //tritave expansion
  //...diamond5Limit,
  //...diamond7Limit
  ...diamond9Limit,
  //...diamond9LimitNo7,
];

//Remove period from target ratios
const periodIndex = targetRatios.findIndex(item => period - item <= 0.000001);
targetRatios.splice(periodIndex, 1);

mainComputation();

//--------------------------
//Computation code
function mainComputation() {
  //get all subset of N notes, with a minimum cents distance
  console.time("calculateSubsets");
  const subsetResult = getEqualTemperamentSubsets(
    numberOfNotes,
    numberOfDivisions,
    noteDistance
  );
  console.timeEnd("calculateSubsets");
  //   console.log(
  //     "initialNumberOfCombinations",
  //     result.initialNumberOfCombinations.toLocaleString()
  //   );
  console.log(
    "numberOfCombinations",
    subsetResult.numberOfCombinations.toLocaleString()
  );

  //Round target ratios to closest equal temperament index
  console.time("getScaleRankInfo");
  let result = [];
  //iterate on each one
  subsetIterator(
    subsetResult.result,
    numberOfNotes,
    subsetResult.numberOfCombinations,
    subset => {
      const scale = subsetToScale(subset, numberOfDivisions, period);

      //compute the rank relative to the target rations an its modes rotation
      const rankInfo = getScaleRankInfo(
        scale,
        targetRatios,
        maxToleranceInCents
      );
      const data = { scale, rankInfo };
      result.push(data);
    }
  );
  console.timeEnd("getScaleRankInfo");

  //Sort by rank
  result = result.sort((a, b) => b.rankInfo.rank - a.rankInfo.rank);

  //Output mapping
  const output = result
    .slice(0, 16)
    //.filter(item => item.rankInfo.closestData.length >= minMatchCount)
    .map((item, resultIndex) => ({
      index: resultIndex,
      rank: item.rankInfo.rank,
      minInterval: item.rankInfo.minInterval,
      scale: item.scale
      //closestData: item.rankInfo.closestData,
      //joinedRotations: item.rankInfo.joinedRotations
    }));
  //console.log(output);
  //const jsonOutput = JSON.stringify(output, undefined, 2)
  const lines = [];
  output.forEach(item => {
    lines.push(`=============`);
    lines.push(`Index: ${item.index}`);
    lines.push(`Rank: ${item.rank}`);
    lines.push(`MinInterval: ${item.minInterval.toFixed(3)}`);
    lines.push(`Scale: ${item.scale.length - 1}`);
    item.scale.forEach(scaleNote => {
      if (scaleNote === 1) {
        return;
      }
      const cents = ratioToCents(scaleNote);
      lines.push(cents.toFixed(3));
    });
  });
  const scaleWorkshopOutput = lines.join("\n");

  fs.writeFile("mos_output.json", scaleWorkshopOutput, err => {
    if (err) {
      throw err;
    }
    console.log("JSON output is saved.");
  });
}

function subsetToScale(subset, numberOfDivisions, period) {
  const scale = [];
  for (let index = 0; index < subset.length; index++) {
    const noteIndex = subset[index];
    scale[index] = getEqualTemperamentNote(
      noteIndex,
      numberOfDivisions,
      period
    );
  }
  scale.push(period);
  return scale;
}

function subsetIterator(
  dataArray,
  numberOfNotes,
  numberOfCombinations,
  callback
) {
  for (let index = 0; index < numberOfCombinations; index++) {
    const start = index * numberOfNotes;
    const end = start + numberOfNotes;
    const subset = dataArray.slice(start, end);
    if (callback !== undefined) {
      callback(subset);
    }
  }
}

function getScaleRankInfo(scale, targetScale, toleranceInCents) {
  const minInterval = getMinInterval(scale);
  const rotations = computeRotations(scale);

  let rank = 0;
  const closestData = [];
  for (let rotation = 0; rotation < rotations.length; rotation++) {
    const rotationScale = rotations[rotation];
    const closestRatios = targetScale.map(targetScaleRatio => {
      const closestRatioInfo = getClosestRatioFromScale(
        targetScaleRatio,
        rotationScale
      );
      return {
        ...closestRatioInfo,
        targetScaleRatio,
        closestScaleRatio: rotationScale[closestRatioInfo.scaleIndex]
      };
    });
    const closestRatiosInfoByTolerance = closestRatios.filter(
      item => item.diffInCents <= toleranceInCents
    );
    let closestDiffInCents = closestRatiosInfoByTolerance.reduce(
      (sum, item) => sum + item.diffInCents,
      0
    );

    //TODO: deduzir numero de dissonancias por rotaçao 16/15 7/5
    const rotationRank =
      closestRatiosInfoByTolerance.length - closestDiffInCents / 1200;
    rank += rotationRank;
    closestData.push(closestRatiosInfoByTolerance);
  }

  let joinedRotations = [].concat.apply([], rotations);
  joinedRotations.sort();
  //joinedRotations.push(period);
  joinedRotations = unique(joinedRotations);

  return {
    rank,
    closestData,
    joinedRotations,
    minInterval
  };
}

File: src3/MicrotonalExplorer/NotesExplorer.cs
using System.Diagnostics;
using System.Text;
using MicrotonalExplorer;

namespace MicrotonalExplorer;
public class NotesExplorer
{
    public static void MainComputation()
    {
        //Parameters:
        var numberOfNotes = 7;
        var numberOfDivisions = 31;
        var period = 2 / 1;
        float maxToleranceInCents = 9;
        var noteDistance = 1; //numberOfDivisions > 43 ? 2 : 1;

        //Remove period from target ratios
        var targetRatios = (
            //Diamond.Diamond5Limit
            //Diamond.Diamond7LimitNo5
            //Diamond.Diamond7Limit
            Diamond.Diamond9Limit
        // new float[]{
        //     16/15f,
        //     10/9f,
        //     8/7f,
        //     7/6f,
        //     6/5f, 
        //     5/4f,
        //     4/3f,
        //     7/5f,
        //     10/7f
        // }
        ).Where(item => period != item).ToArray();

        var targetChords = new List<float[]>{
            // Chord.Overtone5Limit_Inversion0,
            // Chord.Overtone5Limit_Inversion1,
            // Chord.Overtone5Limit_Inversion2,
            // Chord.Undertone5Limit_Inversion0,
            // Chord.Undertone5Limit_Inversion1,
            // Chord.Undertone5Limit_Inversion2,

            Chord.Overtone7Limit_Inversion0,
            Chord.Overtone7Limit_Inversion1,
            Chord.Overtone7Limit_Inversion2,
            Chord.Overtone7Limit_Inversion3,
            Chord.Undertone7Limit_Inversion0,
            Chord.Undertone7Limit_Inversion1,
            Chord.Undertone7Limit_Inversion2,
            Chord.Undertone7Limit_Inversion3,
        };

        var watch = Stopwatch.StartNew();
        var subsetResult = EqualTemperament.GetEqualTemperamentSubsets(numberOfNotes, numberOfDivisions, noteDistance);
        watch.Stop();
        Console.WriteLine($"CalculateSubsets {watch.ElapsedMilliseconds}ms");
        Console.WriteLine($"numberOfCombinations {subsetResult.NumberOfCombinations:N0}");
        //Console.WriteLine(string.Join(',', subsetResult.Result));

        var result = new List<ScaleRankInfoResult>();
        watch = Stopwatch.StartNew();
        //Parallel.For(0, (int)subsetResult.NumberOfCombinations, (index) =>
        for (int index = 0; index < (int)subsetResult.NumberOfCombinations; index++)
        {
            var start = index * numberOfNotes;
            var end = start + numberOfNotes;
            var subset = subsetResult.Result.Skip(start).Take(numberOfNotes);
            var scale = EqualTemperament.ConvertSubsetToScale(subset, numberOfDivisions, period);

            var rankInfo = GetScaleRankInfo(
              scale,
              targetRatios,
              maxToleranceInCents,
              targetChords
            );
            result.Add(rankInfo);
        }
        //});
        Console.WriteLine($"GetScaleRankInfo {watch.ElapsedMilliseconds}ms");

        var orderedResult = result
            .OrderByDescending(item => item.Rank)
            .Take(34);

        SaveOutputFile(orderedResult);
    }


    public static ScaleRankInfoResult GetScaleRankInfo(float[] scale, float[] targetScale, float toleranceInCents, List<float[]> targetChords)
    {
        //Compute the rank relative to the target rations an its modes rotation
        //Considering also a rank addition when some target chords is fully present
        //----------------------------------------------------------------------------

        var minInterval = Operations.GetMinInterval(scale);
        var rotations = Operations.ComputeRotations(scale);

        float rank = 0;
        var closestData = new List<ClosestRationFromScaleResult>();
        for (var rotation = 0; rotation < rotations.Length; rotation++)
        {
            var rotationScale = rotations[rotation];
            float closestDiffInCents = 0;
            int closestCount = 0;
            for (int targetIndex = 0; targetIndex < targetScale.Length; targetIndex++)
            {
                var targetScaleRatio = targetScale[targetIndex];
                var targetResult = Operations.GetClosestRatioFromScale(
                   targetScaleRatio,
                   rotationScale
                 );
                if (targetResult.DiffInCents <= toleranceInCents)
                {
                    closestDiffInCents += targetResult.DiffInCents;
                    closestCount++;
                    closestData.Add(targetResult);
                }
            }

            var rotationRank = closestCount - closestDiffInCents / 1200;
            rank += rotationRank;

            //Full Chord verification
            foreach (var chord in targetChords)
            {
                var chordVerificationResult = Operations.FullMatchTargetRatiosToScale(chord, rotationScale, toleranceInCents);
                if (chordVerificationResult.Any())
                {
                    rank += 7;
                }
            }
        }

        float[] joinedRotations = rotations.SelectMany(item => item).Distinct().Order().ToArray();

        return new ScaleRankInfoResult(
          rank,
          closestData,
          joinedRotations,
          minInterval,
          scale
        );
    }





    static void SaveOutputFile(IEnumerable<ScaleRankInfoResult> result)
    {
        // Write the string array to a new file named "WriteLines.txt".
        using (StreamWriter outputFile = new StreamWriter(Path.Combine("./", "output.txt")))
        {
            foreach (var item in result)
            {
                outputFile.WriteLine("=============");
                //outputFile.WriteLine($"Index: ${item.index}");
                outputFile.WriteLine($"Rank: {item.Rank:0.0000}");
                outputFile.WriteLine($"MinInterval: {item.MinInterval:0.0000}");
                outputFile.WriteLine($"Scale: {item.Scale.Length}");
                foreach (var scaleNote in item.Scale)
                {
                    if (scaleNote == 1)
                    {
                        continue;
                    }
                    var cents = Operations.RatioToCents(scaleNote);
                    outputFile.WriteLine(cents.ToString("0.00"));
                }
            }
        }

    }
}

public record ScaleRankInfoResult(float Rank, IEnumerable<ClosestRationFromScaleResult> ClosestData, float[] JoinedRotations, float MinInterval, float[] Scale);

